Refactoring HiveConnectorTestBase::writeToFile() to use dynamic cast for writer of DWRF or Parquet file format

minhancao opened the pull request:
Refactoring HiveConnectorTestBase::writeToFile() to use dynamic cast for writer of DWRF or Parquet file format. Also refactoring HiveIcebergTest and TableScanTest since they use HiveConnectorTestBase class. 

Related to work in: 
https://github.com/facebookincubator/velox/pull/9973
https://github.com/facebookincubator/velox/pull/10915

Resolves: 
https://github.com/facebookincubator/velox/issues/9171

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -79,19 +87,41 @@ void HiveConnectorTestBase::writeToFile(
+  switch (fileFormat_) {
+#ifdef VELOX_ENABLE_PARQUET
+    case dwio::common::FileFormat::PARQUET: {
+      facebook::velox::parquet::WriterOptions writerOptions;
```
This issue is actually about refactoring and unifying Writer interfaces. See how we made the dwrf::Writer and parquet::Writer classes to all inherit dwio::common::Writer. We need to do similar to the WriterOptions.

minhancao replied:
Edited WriterOptions for DWRF and Parquet to inherit from dwio::common::WriterOptions.

```
struct WriterOptions : dwio::common::WriterOptions
{
...
}
```

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -79,19 +87,41 @@ void HiveConnectorTestBase::writeToFile(
     const std::vector<RowVectorPtr>& vectors,
     std::shared_ptr<dwrf::Config> config,
```
We want to drop the dependency on dwrf, this means we don't want to see the dwrf namesapce being hard coded here. The desired interface is something like this:
```
void writeToFile(
      const std::string& filePath,
      const std::vector<RowVectorPtr>& vectors,
      const TypePtr& schema,
      std::shared_ptr<dwio::common::WriterOptions> options);
```
This allows the user be able to supply custom WriterOptions. the dwrf::WriterOptions contains the dwrf::Config but Parquet doesn't have a counterpart parquet::Config. But they all have WriterOptions and we can unify the interface.

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -79,19 +87,41 @@ void HiveConnectorTestBase::writeToFile(
-  facebook::velox::dwrf::Writer writer{std::move(sink), options};
-  for (size_t i = 0; i < vectors.size(); ++i) {
-    writer.write(vectors[i]);
+  switch (fileFormat_) {
```
@minhancao Can you please try if the following code works? This is more or less what we want it to be. We may need to do some refactor first , including the WriterOptions unification. You may need some other touch ups to make this work.
```
void HiveConnectorTestBase::writeToFile(
    const std::string& filePath,
    const std::vector<RowVectorPtr>& vectors,
    dwio::common::FileFormat fileFormat,
    const TypePtr& schema,
    const facebook::velox::dwio::common::WriterOptions& writerOptions
    ) {
  auto localWriteFile = std::make_unique<LocalWriteFile>(filePath, true, false);
  auto sink = std::make_unique<dwio::common::WriteFileSink>(
      std::move(localWriteFile), filePath);
  auto childPool = rootPool_->addAggregateChild("HiveConnectorTestBase.Writer");
  
  std::unique_ptr<dwio::common::Writer> writer = dwio::common::getWriterFactory(fileFormat)
                    ->createWriter(std::move(sink), writerOptions);
  for (size_t i = 0; i < vectors.size(); ++i) {
    writer->write(vectors[i]);
  }
  writer->close();
```

minhancao replied:
@yingsu00 I tried the following code and it works. However, when I ran `velox_hive_iceberg_test` this is the error that caused all 4 tests to failed, do you maybe have any insight on this error?
```
E20240614 15:23:31.796509 16685645 Exceptions.h:67] Line: /Users/minhancao/velox_2/velox/velox/dwio/parquet/reader/ParquetReader.cpp:163, Function:loadFileMetaData, Expression: strncmp(copy.data() + readSize - 4, "PAR1", 4) == 0 No magic bytes found at end of the Parquet file, Source: RUNTIME, ErrorCode: INVALID_STATE
I20240614 15:23:31.796622 16685645 Task.cpp:1871] Terminating task test_cursor 8 with state Failed after running for 2 ms.
I20240614 15:23:31.796690 16685645 Task.cpp:1135] All drivers (1) finished for task test_cursor 8 after running for 2 ms.
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: No magic bytes found at end of the Parquet file
Retriable: False
Expression: strncmp(copy.data() + readSize - 4, "PAR1", 4) == 0
Context: Split [Hive: /tmp/velox_test_b1U82g 0 - 197922] Task test_cursor 8
Additional Context: Operator: TableScan[0] 0
Function: loadFileMetaData
File: /Users/minhancao/velox_2/velox/velox/dwio/parquet/reader/ParquetReader.cpp
Line: 163
```

```
[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (23277 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] HiveIcebergTest.positionalDeletesSingleBaseFile
[  FAILED  ] HiveIcebergTest.positionalDeletesMultipleBaseFiles
[  FAILED  ] HiveIcebergTest.baseFileMultiplePositionalDeletes
[  FAILED  ] HiveIcebergTest.positionalDeletesMultipleSplits

 4 FAILED TESTS
 ```

minhancao replied:
Fixed velox_hive_iceberg_test error of missing magic bytes at end of Parquet file. Fix was to supply file format to writeToFile() call in IcebergReadTest.cpp that I missed.

minhancao commented on velox/exec/tests/utils/HiveConnectorTestBase.h:
```diff
@@ -207,6 +207,13 @@ class HiveConnectorTestBase : public OperatorTestBase {
+  }
+
+ private:
+  dwio::common::FileFormat fileFormat_{dwio::common::FileFormat::DWRF};
```
Removed

minhancao commented on velox/exec/tests/utils/HiveConnectorTestBase.h:
```diff
@@ -207,6 +207,13 @@ class HiveConnectorTestBase : public OperatorTestBase {
     }
     return assignments;
   }
+
```
Removed `#include "velox/dwio/dwrf/common/Config.h"` from HiveConnectorTestBase.h.

czentgr commented on velox/exec/tests/utils/CMakeLists.txt:
```diff
@@ -50,4 +50,5 @@ target_link_libraries(
   velox_presto_serializer
   velox_functions_prestosql
-  velox_aggregates)
+  velox_aggregates
```
Please order the link libraries alphabetically (also fixing some that are out of order).

czentgr commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -63,35 +63,43 @@ void HiveConnectorTestBase::resetHiveConnector(
+  options.schema = schema;
+
+  // Defaults to DWRF writer if fileFormat is not currently supported yet.
+  if (!hasWriterFactory(fileFormat)) {
```
Why do you need this? You could assume that a factory for the file format exist that is passed down. The test at the beginning wouldn't call this with parquet if the factory was not present in the test.
If the writer type is not present you get an appropriate test error.

czentgr reviewed (commented):
Thanks, a few comments.

czentgr commented on velox/exec/tests/utils/CMakeLists.txt:
```diff
@@ -32,22 +32,23 @@ add_library(
-  velox_exec_test_lib
-  velox_vector_test_lib
-  velox_temp_path
+  velox_aggregates
```
This is not correct. The first one is the target name - and must be `velox_exec_test_lib`. See also the Cmake documentation for `target_link_libraries`.

minhancao replied:
Moved `velox_exec_test_lib` to beginning of the list

czentgr commented on velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:
```diff
@@ -101,54 +101,65 @@ class HiveIcebergTest : public HiveConnectorTestBase {
-    auto task = HiveConnectorTestBase::assertQuery(
-        plan, splits, duckdbSql, numPrefetchSplits);
+    for (dwio::common::FileFormat fileFormat : fileFormats) {
+      // Setting file format in HiveConnectorTestBase.
```
Don't need this comment.

czentgr commented on velox/exec/tests/TableScanTest.cpp:
```diff
@@ -49,6 +50,7 @@ using namespace facebook::velox::exec;
 using namespace facebook::velox::common::test;
 using namespace facebook::velox::exec::test;
 using namespace facebook::velox::tests::utils;
+using namespace facebook::velox::dwrf;
```
This is causing build issues due to ambiguity. See the PR CI pipeline.

minhancao replied:
Thank you! Removed that line and ran `make debug` and it was successful for me.

czentgr commented:
@minhancao please squash your commits and take a look at the failures.

czentgr commented on velox/exec/tests/TableScanTest.cpp:
```diff
@@ -4442,14 +4442,13 @@ TEST_F(TableScanTest, readFlatMapAsStruct) {
-  config->set<const std::vector<uint32_t>>(dwrf::Config::MAP_FLAT_COLS, {0});
-  config->set<const std::vector<std::vector<std::string>>>(
-      dwrf::Config::MAP_FLAT_COLS_STRUCT_KEYS, {keys});
+  dwrf::WriterOptions options;
```
Why are these dwrf options if we want to make this format agnostic?
If this is specifically targeting dwrf maybe we should use a different interface (writeToDwrfFile) and pass dwrf options. or we pass common options and let the common writer factory sort it out from the common options?
It seems a bit more clear to use a specific write function for the cases where it is needed to allow DrwfConfig to be used rather than trying to pollute the common options which can be used for more generic tests.

minhancao replied:
I missed changing this to be `dwio::common::WriterOptions` type, thank you for catching it! Will change accordingly.

czentgr commented on velox/dwio/common/Options.h:
```diff
@@ -640,6 +640,11 @@ struct WriterOptions {
   std::optional<uint8_t> parquetWriteTimestampUnit;
   std::optional<uint8_t> zlibCompressionLevel;
   std::optional<uint8_t> zstdCompressionLevel;
+  std::map<std::string, std::string> configs;
```
What is this for? That seems to be a catch all and I don't really see it used.

minhancao replied:
Removing this since it is not used. I was looking at getDwrfOptions() in /velox/dwio/dwrf/writer/Writer.cpp and it had `std::map<std::string, std::string> configs;` so I was thinking of maybe passing it as part of the common writerOptions but I decided not to go this route

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -61,35 +61,38 @@ void HiveConnectorTestBase::resetHiveConnector(
-  options.config = config;
-  options.schema = schema;
+    dwio::common::WriterOptions options,
+    const TypePtr& schema,
```
See if we can remove this overload if it's only used by the above one.

minhancao replied:
The test TableScanTest::readFlatMapAsStruct needs the schema parameter in TableScanTest.cpp. `writeToFile(file->getPath(), {vector}, options, writeSchema);` so I can't remove this overload

yingsu00 reviewed (commented):
Mostly looks good!

minhancao commented:
@yingsu00 I have rebased with main and fixed merge conflicts, please review when you can!

I've noticed that dwrf configs are passed to HiveConnectorTestBase::writeToFile() quite often in tests so I decided to create my own function for agnostic use, HiveConnectorTestBase::writeToFileCommon(). I suggest maybe also changing writeToFile() to be named writeToFileDwrf() since it seems to only pertain to dwrf file format based from its usage in tests. What do you think?


velox_hive_iceberg_test results:
```
[----------] Global test environment tear-down
[==========] 6 tests from 2 test suites ran. (44251 ms total)
[  PASSED  ] 6 tests.
```

velox_hive_connector_test results:
```
[----------] Global test environment tear-down
[==========] 68 tests from 9 test suites ran. (5355 ms total)
[  PASSED  ] 68 tests.
```

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -103,6 +103,33 @@ void HiveConnectorTestBase::writeToFile(
   writer.close();
 }
 
+void HiveConnectorTestBase::writeToFileCommon(
```
This comment is for line 89
It's still using velox::dwrf::WriterOptions options;
Will you be able to change it to the common one?

yingsu00 replied:
Remove #include "velox/common/file/FileSystems.h"
it's not used

yingsu00 commented on velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:
```diff
@@ -243,11 +253,16 @@ class HiveIcebergTest : public HiveConnectorTestBase {
-          dataVectors,
-          config_,
-          flushPolicyFactory_);
+      if (fileFormat == dwio::common::FileFormat::DWRF) {
```
Similar to writeToFileCommon(), make fileFormat a parameter of writeToFile()

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.h:
```diff
@@ -60,6 +60,17 @@ class HiveConnectorTestBase : public OperatorTestBase {
       const std::function<std::unique_ptr<dwrf::DWRFFlushPolicy>()>&
           flushPolicyFactory = nullptr);
 
```
This file is still dependent on #include "velox/dwio/dwrf/writer/FlushPolicy.h"
Remove the dependency of it.

minhancao replied:
Removing `#include "velox/dwio/dwrf/writer/FlushPolicy.h"` in HiveConnectorTestBase.h requires me to add it to IcebergReadTest.cpp, is that ok?

minhancao commented:
@yingsu00 I have rebased with main to bring in your PR and Pedro's PR, please review when you can, thanks!

I had one error while running the HiveIcebergTest:
```
[ RUN      ] HiveIcebergTest.positionalDeletesMultipleSplits
I20240724 12:04:58.236274 34091896 PeriodicStatsReporter.cpp:69] Starting PeriodicStatsReporter with options allocatorStatsIntervalMs:2000, cacheStatsIntervalMs:2000, arbitratorStatsIntervalMs:2000, spillStatsIntervalMs:2000
I20240724 12:04:58.236331 34091896 HiveConnector.cpp:72] Hive connector test-hive created with maximum of 20000 cached file handles.
I20240724 12:04:58.236330 34092510 PeriodicStatsReporter.cpp:252] Spill memory usage: current[0B] peak[0B]
I20240724 12:04:58.245956 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_TpBDfq,  total size: 41.54KB
I20240724 12:04:58.252242 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_DKzNVD,  total size: 43.27KB
I20240724 12:04:58.258396 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_xxGkPr,  total size: 43.01KB
I20240724 12:04:58.264377 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_Uj7KCA,  total size: 43.39KB
I20240724 12:04:58.270493 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_QlVa5l,  total size: 42.68KB
I20240724 12:04:58.276564 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_CCIW7Z,  total size: 43.38KB
I20240724 12:04:58.282634 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_I3Z1Ao,  total size: 43.06KB
I20240724 12:04:58.288759 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_QRWNe2,  total size: 43.20KB
I20240724 12:04:58.294911 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_EWrb48,  total size: 43.00KB
I20240724 12:04:58.300951 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_WPeKqV,  total size: 43.01KB
I20240724 12:04:58.337361 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_360vuy,  total size: 823B
I20240724 12:04:58.337754 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_duS6np,  total size: 823B
I20240724 12:04:58.338088 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_BOrWxR,  total size: 824B
I20240724 12:04:58.338434 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_m9vFQ5,  total size: 824B
I20240724 12:04:58.338773 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_c2y9Tc,  total size: 824B
I20240724 12:04:58.339118 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_MWMgIO,  total size: 823B
I20240724 12:04:58.339449 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_RWj82r,  total size: 823B
I20240724 12:04:58.339777 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_PfJ2W7,  total size: 823B
I20240724 12:04:58.340096 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_PXIZzM,  total size: 823B
I20240724 12:04:58.341436 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_bPJjyp,  total size: 824B
I20240724 12:04:58.352560 34092514 Task.cpp:1141] All drivers (1) finished for task test_cursor 65 after running for 10 ms.
I20240724 12:04:58.352576 34092514 Task.cpp:1881] Terminating task test_cursor 65 with state Finished after running for 10 ms.
I20240724 12:05:00.241349 34092510 PeriodicStatsReporter.cpp:252] Spill memory usage: current[0B] peak[0B]
I20240724 12:05:00.602698 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_SOBYCO,  total size: 193.28KB
I20240724 12:05:00.611328 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_QdpiJk,  total size: 193.28KB
I20240724 12:05:00.619948 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_ff5cuf,  total size: 193.28KB
I20240724 12:05:00.628422 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_WTEiyn,  total size: 193.28KB
I20240724 12:05:00.637248 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_Q8k11W,  total size: 193.28KB
I20240724 12:05:00.645998 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_qWU5vX,  total size: 193.28KB
I20240724 12:05:00.654503 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_s7jOsF,  total size: 193.28KB
I20240724 12:05:00.662882 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_hGWN9k,  total size: 193.28KB
I20240724 12:05:00.671310 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_qT6nR9,  total size: 193.28KB
I20240724 12:05:00.679775 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_G9HEaf,  total size: 193.28KB
I20240724 12:05:00.716064 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_gol4yd,  total size: 747B
I20240724 12:05:00.716542 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_Jlp9ZY,  total size: 747B
I20240724 12:05:00.716934 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_x61MYj,  total size: 747B
I20240724 12:05:00.717315 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_Ldaf9L,  total size: 747B
I20240724 12:05:00.717693 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_gxRPjB,  total size: 747B
I20240724 12:05:00.718063 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_VKmSvE,  total size: 747B
I20240724 12:05:00.718425 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_CxXyhB,  total size: 747B
I20240724 12:05:00.718778 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_HPaxTP,  total size: 747B
I20240724 12:05:00.719138 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_F6686O,  total size: 747B
I20240724 12:05:00.719501 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_iFNSsl,  total size: 747B
I20240724 12:05:00.732672 34092610 Task.cpp:1141] All drivers (1) finished for task test_cursor 66 after running for 12 ms.
I20240724 12:05:00.732692 34092610 Task.cpp:1881] Terminating task test_cursor 66 with state Finished after running for 12 ms.
I20240724 12:05:02.246464 34092510 PeriodicStatsReporter.cpp:252] Spill memory usage: current[0B] peak[0B]
I20240724 12:05:02.977259 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_HIXdeQ,  total size: 41.54KB
I20240724 12:05:02.983448 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_Co1iql,  total size: 43.27KB
I20240724 12:05:02.989531 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_3U2Kh3,  total size: 43.01KB
I20240724 12:05:02.995604 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_gizLqI,  total size: 43.39KB
I20240724 12:05:03.001678 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_BcGRuK,  total size: 42.68KB
I20240724 12:05:03.007730 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_zkuKP3,  total size: 43.38KB
I20240724 12:05:03.013906 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_bOn7qm,  total size: 43.06KB
I20240724 12:05:03.019982 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_fQfMxn,  total size: 43.20KB
I20240724 12:05:03.026090 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_O9GIgn,  total size: 43.00KB
I20240724 12:05:03.032234 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_jM1lmu,  total size: 43.01KB
I20240724 12:05:03.068727 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_V3grWJ,  total size: 823B
I20240724 12:05:03.069109 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_N1a1c7,  total size: 823B
I20240724 12:05:03.069439 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_IiiGOy,  total size: 823B
I20240724 12:05:03.069805 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_ZknRAz,  total size: 824B
I20240724 12:05:03.070137 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_8VU7Q1,  total size: 823B
I20240724 12:05:03.070472 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_zgZ17d,  total size: 824B
I20240724 12:05:03.070782 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_ulZppl,  total size: 818B
I20240724 12:05:03.071113 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_LFtiPB,  total size: 823B
I20240724 12:05:03.071430 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_W1Qj8A,  total size: 824B
I20240724 12:05:03.071772 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_HXb7Bi,  total size: 824B
I20240724 12:05:03.084453 34092632 Task.cpp:1141] All drivers (1) finished for task test_cursor 67 after running for 12 ms.
I20240724 12:05:03.084478 34092632 Task.cpp:1881] Terminating task test_cursor 67 with state Finished after running for 12 ms.
I20240724 12:05:04.250772 34092510 PeriodicStatsReporter.cpp:252] Spill memory usage: current[0B] peak[0B]
I20240724 12:05:05.337659 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_mFzrk5,  total size: 193.28KB
I20240724 12:05:05.346154 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_25tZKD,  total size: 193.28KB
I20240724 12:05:05.354758 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_5D8vo9,  total size: 193.28KB
I20240724 12:05:05.363242 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_JXYmWs,  total size: 193.28KB
I20240724 12:05:05.371652 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_Tz6OYQ,  total size: 193.28KB
I20240724 12:05:05.380399 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_6DwCPi,  total size: 193.28KB
I20240724 12:05:05.389014 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_GAXUIJ,  total size: 193.28KB
I20240724 12:05:05.397543 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_eAijvC,  total size: 193.28KB
I20240724 12:05:05.405925 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_WL6R9G,  total size: 193.28KB
I20240724 12:05:05.414340 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_jIOHti,  total size: 193.28KB
I20240724 12:05:05.451403 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_DVAzmV,  total size: 747B
I20240724 12:05:05.451829 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_9jpsY1,  total size: 747B
I20240724 12:05:05.452219 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_AtmLbz,  total size: 747B
I20240724 12:05:05.452610 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_IxfNAL,  total size: 747B
I20240724 12:05:05.452991 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_XmiWIY,  total size: 747B
I20240724 12:05:05.453364 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_JtfXTH,  total size: 747B
I20240724 12:05:05.453716 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_qgsZhB,  total size: 747B
I20240724 12:05:05.454093 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_3HaSAm,  total size: 747B
I20240724 12:05:05.454448 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_w1nUpu,  total size: 747B
I20240724 12:05:05.454800 34091896 FileSink.cpp:125] closing file: /tmp/velox_test_9jgJ1x,  total size: 747B
E20240724 12:05:05.475178 34092720 Exceptions.h:67] Line: /Users/minhancao/velox_2/velox/velox/common/file/File.cpp:118, Function:LocalReadFile, Expression:  open failure in LocalReadFile constructor, -1 /tmp/velox_test_9jpsY1 Too many open files., Source: RUNTIME, ErrorCode: INVALID_STATE
I20240724 12:05:05.475392 34092720 Task.cpp:1881] Terminating task test_cursor 68 with state Failed after running for 20 ms.
I20240724 12:05:05.475462 34092720 Task.cpp:1141] All drivers (1) finished for task test_cursor 68 after running for 20 ms.
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: open failure in LocalReadFile constructor, -1 /tmp/velox_test_9jpsY1 Too many open files.
Retriable: False
Context: Split [Hive: /tmp/velox_test_GAXUIJ 0 - 197922] Task test_cursor 68
Additional Context: Operator: TableScan[0] 0
Function: LocalReadFile
File: /Users/minhancao/velox_2/velox/velox/common/file/File.cpp
Line: 118
```
Do you maybe have any insight for this error?


Ying's PR:
https://github.com/facebookincubator/velox/pull/10505/

Pedro's PRs:
https://github.com/facebookincubator/velox/pull/10470/
https://github.com/facebookincubator/velox/pull/10520/ (to fix build errors caused by previous PR)

minhancao commented:
@yingsu00 @czentgr I noticed from Pedro's recent PR https://github.com/facebookincubator/velox/pull/10470, he moved the parquet specific option properties in dwio/common/Options.h to only be in the parquet::WriterOptions in dwio/parquet/writer/Writer.h. In parquet/writer/Writer.cpp, he also rewrote the createWriter() function to be:
```
std::unique_ptr<dwio::common::Writer> ParquetWriterFactory::createWriter(
    std::unique_ptr<dwio::common::FileSink> sink,
    const std::shared_ptr<dwio::common::WriterOptions>& options) {
  auto parquetOptions =
      std::dynamic_pointer_cast<parquet::WriterOptions>(options);
  VELOX_CHECK_NOT_NULL(
      parquetOptions,
      "Parquet writer factory expected a Parquet WriterOptions object.");
  return std::make_unique<Writer>(
      std::move(sink), *parquetOptions, asRowType(options->schema));
}
```

The current route that my PR is taking is providing all options of DWRF and Parquet into dwio/common/Options.h and populating it accordingly based on the file format for its respective writers. I have edited the code above to go to my route when the dynamic cast fails.
```
std::unique_ptr<dwio::common::Writer> ParquetWriterFactory::createWriter(
    std::unique_ptr<dwio::common::FileSink> sink,
    const std::shared_ptr<dwio::common::WriterOptions>& options) {
  auto parquetOptions =
      std::dynamic_pointer_cast<parquet::WriterOptions>(options);

  if (parquetOptions == nullptr) {
    auto parquetOptionsFromCommon = getParquetOptions(*options);
    return std::make_unique<Writer>(
        std::move(sink), parquetOptionsFromCommon, asRowType(options->schema));
  }

  return std::make_unique<Writer>(
      std::move(sink), *parquetOptions, asRowType(options->schema));
}
```

Is this ok? Or is there a better way to approach this?

It seems to me that we are in conflict now of whether we should take the route of providing all file format options into dwio/common/Options.h and populating it accordingly OR providing the specific file format WriterOptions at the beginning in the test or wherever it's called and dynamically casting it when it is received to its respective file format type.

minhancao commented:
Hi @pedroerp! I saw from your recent PR https://github.com/facebookincubator/velox/pull/10470 you moved the Parquet specific options from dwio/common/Options.h to be in the specific Parquet writer options. Do you think the DWRF options that are in dwio/common/Options.h should be moved to the specific DWRF writer options as well and refactor accordingly? 

This PR was meant to refactor HiveConnectorTestBase::writeToFile() to be able to be agnostic and use DWRF or Parquet writer depending on the file format that is passed in.

pedroerp commented:
> Do you think the DWRF options that are in dwio/common/Options.h should be moved to the specific DWRF writer options as well and refactor accordingly?

Hi @minhancao, yes, for sure. I have a draft PR for that, but it turns out to be a bit more refactoring that I expected, but I'm planning to get back at it next week.

minhancao commented:
@pedroerp Ok awesome! Can you please link me the PR whenever it's ready?

yingsu00 commented:
@minhancao Can you please squash the commits into one?

yingsu00 commented on velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:
```diff
@@ -19,6 +19,7 @@
 #include "velox/connectors/hive/iceberg/IcebergDeleteFile.h"
 #include "velox/connectors/hive/iceberg/IcebergMetadataColumns.h"
 #include "velox/connectors/hive/iceberg/IcebergSplit.h"
+#include "velox/dwio/dwrf/writer/FlushPolicy.h"
```
Why is it still dependent on DWRF? Can you please remove the dependency? You may need to do some more refactoring to use the dwio/common/ flush policy.

minhancao replied:
There's this piece of code:
```
HiveIcebergTest()
      : config_{std::make_shared<facebook::velox::dwrf::Config>()} {
    // Make the writers flush per batch so that we can create non-aligned
    // RowGroups between the base data files and delete files
    flushPolicyFactory_ = []() {
      return std::make_unique<dwrf::LambdaFlushPolicy>([]() { return true; });
    };
  }
  ```
If I remove `#include "velox/dwio/dwrf/writer/FlushPolicy.h"` then it won't work...

czentgr replied:
You can argue that the `LambdaFlushPolicy` should be a format independent `FlushPolicy` implementation and be moved to dwio::common. `FlushPolicy` is an interface already in common.

You can do this as a separate PR. Check if the Parquet also implements the same LambdaFlushPolicy. The thing about it is that nothing from the format is taken into account. Only the lambda decides so it is common among the formats.

minhancao replied:
Looking at DWRF LambdaFlushPolicy in dwrf/FlushPolicy.h, it inherits a function from the DWRFFlushPolicy class called shouldFlushDictionary(). 

Also looking at Parquet LambdaFlushPolicy in parquet/writer/Writer.h, it inherits 2 variables from Parquet's DefaultFlushPolicy class that are not in the common FlushPolicy:
```
uint64_t rowsInRowGroup,
int64_t bytesInRowGroup,
```

Looks like since LambdaFlushPolicy inherits some properties/functions from a FlushPolicy class that's specifically for its respective file formats (dwrf and parquet), we shouldn't move the LambdaFlushPolicy to be in the common FlushPolicy.h... @czentgr What do you think?

czentgr replied:
I see, the FlushPolicy's are format specific for certain function calls. And if they derive from a different class then the usage wouldn't work anymore. You would need to build a class that inherits from both, the common LambdaFlushPolicy and the format specific FlushPolicy. But this type of polymorphic inheritance might not be desired (could get messy). Doesn't look like there is an easy solution to this unless the common LambdaFlushPolicy combines the format specific API/members.

minhancao replied:
@yingsu00 Can we maybe get your thoughts on how we can refactor accordingly for this?

yingsu00 replied:
@minhancao 
I think we can just use the following structure:
```
             dwio::common::FlushPolicy
            /                             \
DWRFFlushPolicy          ParquetFlushPolicy (original DefaultFlushPolicy)
          |                                       |
dwrf::LambdaFlushPolicy            parquet::LambdaFlushPolicy
```

minhancao replied:
https://github.com/facebookincubator/velox/pull/8864

yingsu00 replied:
@minhancao So these recent changes are already made by someone else and would save you some effort. Then why is this file still dependent  on dwrf? Please, let it depend on dwio::common, not dwrf.

minhancao replied:
@yingsu00 
I don't think it can depend on dwio::common because LambdaFlushPolicy is either respectively in DWRF or Parquet's flush policy. LambdaFlushPolicy is not in dwio::common::FlushPolicy.

Also, the flow now that I am implementing is you instantiate the respective file format's writer options in the test and then pass it to HiveConnectorTestBase::writeToFile() function where there will be dynamic casting of the WriterOptions based on the FileFormat passed into the function.

This PR's implementation is based on this PR https://github.com/facebookincubator/velox/pull/10915 that was merged in last month.

yingsu00 commented on velox/exec/tests/utils/HiveConnectorTestBase.h:
```diff
@@ -20,7 +20,6 @@
 #include "velox/connectors/hive/HiveDataSink.h"
 #include "velox/connectors/hive/TableHandle.h"
 #include "velox/dwio/dwrf/common/Config.h"
```
Here it's still dependent on dwrf. Please remove it. Do proper refactor if you need.

minhancao replied:
Removed, build was successful so no need to refactor.

aditi-pandit commented on velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:
```diff
@@ -170,66 +171,76 @@ class HiveIcebergTest : public HiveConnectorTestBase {
+          if (deleteFileContent.count(baseFileName) != 0) {
+            // If this delete file contains rows for the target base file, then
+            // add it to the split
+            auto deleteFilePath =
```
This variable is not needed. The definition can be used in-place.

minhancao replied:
Changing to:
```
IcebergDeleteFile deleteFile(
                FileContent::kPositionalDeletes,
                deleteFilePaths[deleteFileName].second->getPath(),
                fileFormat,
                deleteFilePaths[deleteFileName].first,
                testing::internal::GetFileSize(
                    std::fopen(deleteFilePath.c_str(), "r")));
```

aditi-pandit commented on velox/dwio/common/Options.h:
```diff
@@ -36,6 +36,7 @@
 #include "velox/dwio/common/encryption/Encryption.h"
 #include "velox/type/Timestamp.h"
 #include "velox/type/tz/TimeZoneMap.h"
+#include "velox/vector/arrow/Bridge.h"
```
Nit : Why is this include needed ?

minhancao replied:
It's for TimestampUnit type on line 631:
```
// Parquet specific options.
  std::optional<TimestampUnit> parquetWriteTimestampUnit;
 ```

czentgr replied:
Yes, but the type from the config is read as `uint8_t` - it is represented externally in the HiveConfig and read by `readTimestampUnit` from the config? So this can be used and then get converted to the type for the parquet writer later? That way we don't leak the vector/arrow header into common headers.

minhancao replied:
Removed include

aditi-pandit reviewed (commented):
Thanks @minhancao. Some minor comments.,

minhancao commented:
Hi @pedroerp! Just wanted to check in if you were working on a PR for the DWRF options that are in dwio/common/Options.h to be moved to the specific DWRF writer options just like how you did with Parquet options? If not, I can take up this work.

pedroerp commented:
> Hi @pedroerp! Just wanted to check in if you were working on a PR for the DWRF options that are in dwio/common/Options.h to be moved to the specific DWRF writer options just like how you did with Parquet options? If not, I can take up this work.

@minhancao I haven't had time to look into that yet, but I'm happy to help with code review if this is something you would like to work on. @Yuhta has a lot more context on that code though. 

The one tricky part is to make sure we maintain the same behavior and default values; I remember seeing that there a couple of different places where these are defined for DWRF today.

minhancao commented:
Looks like DWRF and Parquet writer options were both refactored to be using dynamic_cast now thanks to this PR 2 weeks ago:
https://github.com/facebookincubator/velox/pull/10915

minhancao commented:
Refactored HiveConnectorTestBase::writeToFile() and tests using it based on https://github.com/facebookincubator/velox/pull/10915.

czentgr commented on velox/exec/tests/utils/HiveConnectorTestBase.h:
```diff
@@ -43,23 +41,32 @@ class HiveConnectorTestBase : public OperatorTestBase {
+  void writeToFile(
+      const std::string& filePath,
+      const RowVectorPtr vector,
+      const dwio::common::FileFormat& fileFormat =
```
For the fileFormat the parameter doesn't need to be a const & as it is an enum (so it similar to numeric types).

minhancao replied:
Changing to `dwio::common::FileFormat fileFormat`

minhancao commented on velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:
```diff
@@ -243,11 +258,44 @@ class HiveIcebergTest : public HiveConnectorTestBase {
+        case dwio::common::FileFormat::PARQUET: {
+          auto parquetOptions = std::make_shared<parquet::WriterOptions>();
+          parquetOptions->flushPolicyFactory = []() {
+            return std::make_unique<parquet::LambdaFlushPolicy>(
```
Adding #ifdef VELOX_ENABLE_PARQUET to this case statement

minhancao commented on velox/exec/tests/utils/HiveConnectorTestBase.cpp:
```diff
@@ -64,46 +64,70 @@ void HiveConnectorTestBase::resetHiveConnector(
+      break;
+#endif
+    default:
+      writerOptions = std::dynamic_pointer_cast<dwrf::WriterOptions>(options);
```
Replaced with `VELOX_NYI("Fileformat not supported: {}.", dwio::common::toString(fileFormat));`

minhancao replied:
@czentgr I tried the above and it caused [333 - velox_exec_test (Subprocess aborted) ](https://github.com/facebookincubator/velox/actions/runs/11167598882/job/31044148858?pr=10150)to fail... I reverted it back and it's working fine now...

minhancao replied:
Turns out the failure is due to this flaky test MultiFragmentTest.exchangeStatsOnFailure: https://github.com/facebookincubator/velox/issues/10899

Reverting back to VELOX_NYI implementation.

majetideepak commented:
CC: @svm1
